File: javascript/src/filter.ts
/**
 * Visibility filters.
 *
 * A {@link Filter} controls which changes a document renders. Changes the
 * filter rejects are still stored in the document and synced to peers, but
 * they do not appear in the rendered state. Setting the filter rebuilds the
 * op-set index in one pass; reads stay at full speed afterwards.
 *
 * A filter is a small set of {@link Rule}s, each scoped to one of three
 * places:
 *
 * - the document **default** ({@link Filter.default}),
 * - a particular **author** ({@link Filter.authors}, keyed by the hex
 *   author identifier — see {@link getAuthor}), or
 * - a particular **actor** ({@link Filter.actors}, keyed by an
 *   {@link Actor} ID).
 *
 * Resolution is most-specific-wins: an actor rule overrides an author rule,
 * which overrides the default.
 *
 * # Examples
 *
 * Hide every change made by an author after a known set of heads. This is
 * the "revoke writes by Alice after we last trusted her" pattern:
 *
 * ```ts
 * import * as Automerge from "@automerge/automerge"
 *
 * doc = Automerge.updateFilter(doc, f => {
 *   f.authors ??= {}
 *   f.authors[alice] = Automerge.allowUpTo(trustedHeads)
 * })
 * ```
 *
 * Render only a validated prefix, but keep accepting changes from the local
 * actor (the schema-validation pattern):
 *
 * ```ts
 * doc = Automerge.setFilter(doc, {
 *   default: Automerge.allowUpTo(validatedHeads),
 *   actors: { [Automerge.getActorId(doc)]: Automerge.allow },
 * })
 * ```
 *
 * Drop the rule for one author without disturbing the rest of the filter:
 *
 * ```ts
 * doc = Automerge.updateFilter(doc, f => { delete f.authors?.[alice] })
 * ```
 *
 * Clear every rule:
 *
 * ```ts
 * doc = Automerge.setFilter(doc, {})
 * ```
 *
 * @module
 */

import { _state } from "./internal_state.js"
import type { ApplyOptions } from "./implementation.js"
import { progressDocument } from "./implementation.js"
import type { Doc } from "./types.js"
import type { Filter, Heads, Rule } from "./wasm_types.js"

export type { Filter, Rule } from "./wasm_types.js"

/**
 * The {@link Rule} that accepts every matching change.
 *
 * Equivalent to the literal `"allow"` — exported so that idiomatic builder
 * code reads naturally:
 *
 * ```ts
 * doc = Automerge.updateFilter(doc, f => {
 *   f.actors ??= {}
 *   f.actors[peer] = Automerge.allow
 * })
 * ```
 */
export const allow: Rule = "allow"

/**
 * The {@link Rule} that rejects every matching change.
 *
 * Equivalent to the literal `"deny"`. See {@link allow} for the rationale
 * behind exporting both as named constants.
 */
export const deny: Rule = "deny"

/**
 * Build the {@link Rule} that accepts only changes which are ancestors of
 * `heads`. Heads do not need to be present in the document at the time the
 * rule is set: a head referenced here that arrives later is honoured as
 * soon as it lands.
 */
export function allowUpTo(heads: Heads): Rule {
  return { allowUpTo: [...heads] }
}

/**
 * Get the visibility filter currently in effect for `doc`.
 *
 * The returned object is a fresh copy; mutating it does not affect the
 * document. Use {@link setFilter} or {@link updateFilter} to install a new
 * filter.
 */
export function getFilter<T>(doc: Doc<T>): Filter {
  return _state(doc).handle.getFilter()
}

/**
 * Replace the visibility filter for `doc` and return the resulting
 * document.
 *
 * Changes the filter accepts are rendered; rejected changes remain stored
 * and continue to sync, they just do not appear in the rendered state.
 *
 * Pass `{}` to clear every rule.
 *
 * @param doc - The document to update.
 * @param filter - The new filter. Any field omitted is treated as empty
 *     (no rules in that scope, default `"allow"`).
 * @param opts - Standard {@link ApplyOptions}: pass a `patchCallback` to
 *     observe the patches caused by changes that newly became visible or
 *     newly became hidden.
 * @returns A new document handle reflecting the new filter.
 */
export function setFilter<T>(
  doc: Doc<T>,
  filter: Filter,
  opts?: ApplyOptions<T>,
): Doc<T> {
  const state = _state(doc)
  if (state.heads) {
    throw new RangeError(
      "Attempting to change an outdated document. Use Automerge.clone() if you wish to make a writable copy.",
    )
  }
  const heads = state.handle.getHeads()
  state.handle.setFilter(filter)
  return progressDocument(
    doc,
    "setFilter",
    heads,
    opts?.patchCallback ?? state.patchCallback,
  )
}

/**
 * Read-modify-write helper around {@link setFilter}.
 *
 * `fn` is called with a mutable copy of the current filter and may add,
 * remove, or change rules in place. Once it returns the modified filter is
 * installed.
 *
 * This is the most ergonomic way to add or remove a single rule:
 *
 * ```ts
 * doc = Automerge.updateFilter(doc, f => {
 *   f.authors ??= {}
 *   f.authors[alice] = Automerge.allowUpTo(trustedHeads)
 * })
 * ```
 *
 * `opts` matches {@link setFilter}.
 */
export function updateFilter<T>(
  doc: Doc<T>,
  fn: (filter: Filter) => void,
  opts?: ApplyOptions<T>,
): Doc<T> {
  const filter = getFilter(doc)
  fn(filter)
  return setFilter(doc, filter, opts)
}

File: rust/automerge/src/filter.rs
//! Persistent visibility filters.
//!
//! A [`Filter`] tells the document which changes should be rendered. Changes
//! that are rejected by the active filter are still ingested and synced to
//! peers — only their effect on the rendered state is suppressed. Changing
//! the filter rebuilds the op-set index in one O(n) pass; afterwards reads
//! stay at full speed because the index itself is the filtered view.
//!
//! Each rule applies to a *scope* (the document default, an [`Author`], or
//! an [`ActorId`]) and selects one of three behaviours:
//!
//! * [`Rule::Allow`] — accept every matching change.
//! * [`Rule::AllowUpTo`] — accept matching changes that are ancestors of the
//!   given heads. Heads do not have to be known to the document yet; a rule
//!   referencing an unseen head takes effect once the change arrives.
//! * [`Rule::Deny`] — reject every matching change.
//!
//! Resolution for a given actor is *most-specific-wins*: an [`ActorId`] rule
//! overrides an [`Author`] rule, which overrides the document default.
//!
//! # Examples
//!
//! Reject every change made by an author after a known set of heads:
//!
//! ```ignore
//! let filter = Filter::default()
//!     .with_author(alice.clone(), Rule::AllowUpTo { heads: known_good });
//! doc.set_filter(filter);
//! ```
//!
//! Render only a validated prefix, but keep accepting changes from the
//! local actor and from one trusted peer:
//!
//! ```ignore
//! let filter = Filter {
//!     default: Rule::AllowUpTo { heads: validated_heads },
//!     ..Filter::default()
//! }
//! .with_actor(local_actor.clone(), Rule::Allow)
//! .with_actor(peer_actor.clone(), Rule::Allow);
//! doc.set_filter(filter);
//! ```

use std::collections::BTreeMap;

use crate::types::{ActorId, Author};
use crate::ChangeHash;

/// What the filter does with changes that fall in a rule's scope.
#[derive(Clone, Debug, PartialEq, Eq, Default)]
pub enum Rule {
    /// Accept every matching change.
    #[default]
    Allow,
    /// Accept matching changes only if they are ancestors of `heads`.
    ///
    /// Heads do not need to be present in the document at the time the rule
    /// is set: a head referenced here that arrives later is honoured as
    /// soon as it lands.
    AllowUpTo { heads: Vec<ChangeHash> },
    /// Reject every matching change.
    Deny,
}

/// A persistent visibility filter. See the [module docs][self] for an
/// overview.
///
/// `Filter` is a pure rule set — the document holds onto it across saves
/// and loads, and resolves it into a per-actor mask whenever it changes.
/// Construct one with [`Filter::default`] and the `with_*` builders, or by
/// populating the public fields directly.
#[derive(Clone, Debug, PartialEq, Eq, Default)]
pub struct Filter {
    /// Applied to changes that don't match a more specific rule.
    pub default: Rule,
    /// Per-author overrides. Override the [`Filter::default`] rule for any
    /// actor whose change carries this author.
    pub authors: BTreeMap<Author, Rule>,
    /// Per-actor overrides. Override both [`Filter::default`] and any
    /// matching [`Filter::authors`] rule.
    pub actors: BTreeMap<ActorId, Rule>,
}

impl Filter {
    /// An empty filter — every change is accepted.
    pub fn new() -> Self {
        Self::default()
    }

    /// Replace the default rule.
    pub fn with_default(mut self, rule: Rule) -> Self {
        self.default = rule;
        self
    }

    /// Set or replace the rule for `author`.
    pub fn with_author(mut self, author: Author, rule: Rule) -> Self {
        self.authors.insert(author, rule);
        self
    }

    /// Set or replace the rule for `actor`.
    pub fn with_actor(mut self, actor: ActorId, rule: Rule) -> Self {
        self.actors.insert(actor, rule);
        self
    }

    /// Whether the filter is identical to [`Filter::default`].
    ///
    /// When this is true the filter accepts every change, and read paths
    /// can skip filter-related work entirely.
    pub(crate) fn is_noop(&self) -> bool {
        matches!(self.default, Rule::Allow) && self.authors.is_empty() && self.actors.is_empty()
    }
}

/// The rule that applies to a particular actor, after most-specific-wins
/// resolution. Borrows the rule out of the filter to avoid cloning the
/// (potentially long) head list inside `AllowUpTo`.
pub(crate) enum ResolvedRule<'a> {
    Allow,
    AllowUpTo(&'a [ChangeHash]),
    Deny,
}

impl<'a> From<&'a Rule> for ResolvedRule<'a> {
    fn from(r: &'a Rule) -> Self {
        match r {
            Rule::Allow => ResolvedRule::Allow,
            Rule::AllowUpTo { heads } => ResolvedRule::AllowUpTo(heads.as_slice()),
            Rule::Deny => ResolvedRule::Deny,
        }
    }
}

impl Filter {
    /// Every head referenced by any rule. Used by the change graph to
    /// track which rules are waiting on changes that haven't arrived yet.
    pub(crate) fn referenced_heads(&self) -> impl Iterator<Item = &ChangeHash> {
        fn heads_of(rule: &Rule) -> &[ChangeHash] {
            match rule {
                Rule::AllowUpTo { heads } => heads.as_slice(),
                _ => &[],
            }
        }
        heads_of(&self.default)
            .iter()
            .chain(self.authors.values().flat_map(|r| heads_of(r).iter()))
            .chain(self.actors.values().flat_map(|r| heads_of(r).iter()))
    }
}
